pop-api/examples/fungibles/lib.rs

```diff
@@ -1,108 +1,261 @@
+			let allowance = self.allowance(caller, spender);
+			self.env().emit_event(Approval { owner: caller, spender, value: allowance });
+			Ok(())
 		}
 
+		/// Decreases by `value` the allowance granted to `spender` by the caller.
 		#[ink(message)]
-		pub fn decrease_allowance(
+		fn decrease_allowance(
 			&mut self,
-			token: TokenId,
 			spender: AccountId,
 			value: Balance,
-		) -> Result<()> {
-			api::decrease_allowance(token, spender, value)
+		) -> Result<(), PSP22Error> {
+			let caller = self.env().caller();
+			// No-op if the caller and `spender` is the same address or `value` is zero, returns
+			// success and no events are emitted.
+			if caller == spender || value == 0 {
+				return Ok(());
+			}
+			// Reverts with `InsufficientAllowance` if `spender` and the caller are different
+			// addresses and the `value` exceeds the allowance granted by the caller to
+			// `spender`.
+			let allowance = self.allowance(caller, spender);
+			if allowance < value {
+				return Err(PSP22Error::InsufficientAllowance);
+			}
+
+			api::decrease_allowance(self.id, spender, value).map_err(PSP22Error::from)?;
+			self.env().emit_event(Approval {
+				owner: caller,
+				spender,
+				value: allowance.saturating_sub(value),
+			});
+			Ok(())
+		}
+	}
+
+	impl Psp22Metadata for Fungibles {
+		/// Returns the token name.
+		#[ink(message)]
+		fn token_name(&self) -> Option<String> {
+			api::token_name(self.id)
+				.ok()
+				.filter(|v| !v.is_empty())
+				.and_then(|v| String::from_utf8(v).ok())
 		}
 
+		/// Returns the token symbol.
 		#[ink(message)]
-		pub fn token_name(&self, token: TokenId) -> Result<Vec<u8>> {
-			api::token_name(token)
+		fn token_symbol(&self) -> Option<String> {
+			api::token_symbol(self.id)
+				.ok()
+				.filter(|v| !v.is_empty())
+				.and_then(|v| String::from_utf8(v).ok())
 		}
 
+		/// Returns the token decimals.
 		#[ink(message)]
-		pub fn token_symbol(&self, token: TokenId) -> Result<Vec<u8>> {
-			api::token_symbol(token)
+		fn token_decimals(&self) -> u8 {
+			api::token_decimals(self.id).unwrap_or_default()
 		}
+	}
+
+	impl Psp22Mintable for Fungibles {
+		/// Mints `value` tokens to the senders account.
+		#[ink(message)]
+		fn mint(&mut self, account: AccountId, value: Balance) -> Result<(), PSP22Error> {
+			if value == 0 {
+				return Ok(());
+			}
+			api::mint(self.id, account, value).map_err(PSP22Error::from)?;
+			self.env().emit_event(Transfer { from: None, to: Some(account), value });
+			Ok(())
+		}
+	}
 
+	impl Psp22Burnable for Fungibles {
+		/// Burns `value` tokens from the senders account.
 		#[ink(message)]
-		pub fn token_decimals(&self, token: TokenId) -> Result<u8> {
-			api::token_decimals(token)
+		fn burn(&mut self, account: AccountId, value: Balance) -> Result<(), PSP22Error> {
+			if value == 0 {
+				return Ok(());
+			}
+			let balance = self.balance_of(account);
+			if balance < value {
+				return Err(PSP22Error::InsufficientBalance);
```

Comment:
`InsufficientBalance` case is not returned in the `burn` pallet api if the `value` exceeds the minted value. In `decrease_balance` method of `pallet-assets`, it is also `saturating_sub`.

Reply:
Agree to keep it as it is.

Reply:
Would be cheaper in the pallet though. 

The pallet will also be compliant then, not sure how useful that is.

Something to decide on tomorrow

Reply:
Separated PR: https://github.com/r0gue-io/pop-node/pull/311